src/App.tsx

```diff
@@ -1,24 +1,220 @@
 
-export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+export const App: FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isInputDisabled, setIsInputDisabled] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [filterMethod, setFilterStatus] = useState(FilteringMethod.ALL);
+  const [loadingTodosIDs, setLoadingTodosIDs] = useState<number[]>([]); // problem is here
+  const [textFieldValue, setTextFieldValue] = useState('');
+  const [errorMessage, setErrorMessage] = useState(ErrorMessage.NONE);
+
+  useEffect(() => {
+    (async () => {
+      try {
+        const todosFromServer = await getTodos(USER_ID);
+
+        setTodos(todosFromServer);
+      } catch {
+        setErrorMessage(ErrorMessage.LOAD);
+
+        setTimeout(() => {
+          setErrorMessage(ErrorMessage.NONE);
+        }, 3000);
+      }
+    })();
+  }, []);
+
+  const showErrorMessage = useCallback((message: ErrorMessage) => {
+    setErrorMessage(message);
+    setTimeout(() => {
+      setErrorMessage(ErrorMessage.NONE);
+    }, 3000);
+  }, []);
+
+  const setTodoLoading = useCallback((id: number) => {
+    setLoadingTodosIDs(prevState => [...prevState, id]);
+  }, []);
+
+  const unsetTodoLoading = useCallback((id: number) => {
+    setLoadingTodosIDs(prevState => prevState.filter(num => num !== id));
+  }, []);
+
+  const addTodo = useCallback(async () => {
+    setIsInputDisabled(true);
+
+    const newTodo = {
+      title: textFieldValue,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({
+      ...newTodo,
+      id: 0,
+    });
+
+    try {
+      const resultingTodo = await postTodo(newTodo);
+
+      setTodos(prev => [
+        ...prev,
+        resultingTodo,
+      ]);
+      setTextFieldValue('');
+    } catch {
+      showErrorMessage(ErrorMessage.ADD);
+    } finally {
+      setIsInputDisabled(false);
+      setTempTodo(null);
+    }
+  }, [textFieldValue]);
+
+  const deleteTodo = useCallback(async (id: number) => {
+    setTodoLoading(id);
+
+    try {
+      await removeTodo(id);
+
+      const newTodosList = todos.filter(todo => todo.id !== id);
+
+      setTodos(() => newTodosList);
+    } catch {
+      showErrorMessage(ErrorMessage.DELETE);
+    } finally {
+      unsetTodoLoading(id);
+    }
+  }, [todos]);
+
+  const deleteAllCompleted = useCallback(async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    completedTodos.forEach(todo => {
+      setTodoLoading(todo.id);
+    });
+
+    try {
+      await Promise.all(completedTodos.map(todo => removeTodo(todo.id)));
+
+      const uncompletedTodos = todos.filter(todo => !todo.completed);
+
+      setTodos(uncompletedTodos);
+    } catch {
+      showErrorMessage(ErrorMessage.DELETE);
+    } finally {
+      setLoadingTodosIDs([]);
+    }
+  }, [todos]);
+
+  const updateTodo = useCallback(async (id: number, data: object) => {
```

Comment:
object?

Reply:
let's use proper type here, it can't be any random object

Reply:
since you updating only part of Todo object here, I guess utility type [Partial](https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype) - is a good option here